Database triggers don't work for batch inserts

* pro `answers` bud nemuzeme pouzit trigger, nebo vyhodit batch 
```
05/03/2021 15:13:11 [DEBUG] [ce4b3595-7f7c-7f53-efa2-55db0d3f07a6] Exposed - INSERT INTO patients (email, first_name, id, insurance_company, last_name, personal_number, phone_number, updated) VALUES ('john@doe.com', 'John', 'e292d01d-cf08-49ee-b383-ae79255f0cfe', 'RBP', 'Doe', '7401040020', '+420123456789', NULL)
05/03/2021 15:13:11 [ERROR] [ce4b3595-7f7c-7f53-efa2-55db0d3f07a6] b.mild.covid.vaxx.ExceptionHandler - Exception occurred in the application: Can't add a new batch because columns: answers.created don't have default values. DB defaults don't support in batch inserts
org.jetbrains.exposed.sql.statements.BatchDataInconsistentException: Can't add a new batch because columns: answers.created don't have default values. DB defaults don't support in batch inserts
```

```bash
curl 'https://covid-vaxx.stg.mild.blue/api/patient' \
  -H 'content-type: application/json' \
  --data-raw '{"answers":[{"questionId":"9a5587a1-dc43-49f3-9847-b736127c9e39","value":"true"},{"questionId":"f74ebe1e-ef94-4af0-963d-97ffab086b6b","value":"true"},{"questionId":"f68d221d-27a1-4c81-bf45-07b1f0290e15","value":"false"},{"questionId":"f9c99047-0f44-4dfe-9964-71274a7af5e9","value":"false"},{"questionId":"f5cf0689-a4d7-4c42-8107-6eaedca88a93","value":"true"},{"questionId":"7b02b12a-abb4-45d3-8bf4-0b074e445f37","value":"false"},{"questionId":"112f5fbd-cde2-4fe9-8cab-f5b4fff57296","value":"true"},{"questionId":"f4ca8d25-faaa-4b2f-abc2-3d7a8702d4a3","value":"true"}],"confirmation":{"covid19VaccinationAgreement":true,"healthStateDisclosureConfirmation":true},"firstName":"John","lastName":"Doe","personalNumber":"7401040020","insuranceCompany":"RBP","phoneNumber":"+420123456789","email":"john@doe.com"}' \
  --compressed
```

--- a/backend/src/main/kotlin/blue/mild/covid/vaxx/auth/JwtService.kt
+++ b/backend/src/main/kotlin/blue/mild/covid/vaxx/auth/JwtService.kt
@@ -2,7 +2,7 @@ package blue.mild.covid.vaxx.auth
 
 import blue.mild.covid.vaxx.dao.UserRole
 import blue.mild.covid.vaxx.dto.JwtConfigurationDto
-import blue.mild.covid.vaxx.dto.response.JwtResponseDtoOut
+import blue.mild.covid.vaxx.dto.response.BearerTokenDtoOut
 import com.auth0.jwt.JWT
 import com.auth0.jwt.JWTCreator
 import com.auth0.jwt.JWTVerifier
@@ -38,7 +38,7 @@ class JwtService(
     /**
      * Generate and sign JWT for given principal.¬
      */
-    fun generateToken(principal: UserPrincipal): JwtResponseDtoOut {
+    fun generateToken(principal: UserPrincipal): BearerTokenDtoOut {
         val builder = JWT.create()
             .withIssuer(jwtConfiguration.issuer)
             .withAudience(jwtConfiguration.audience)
@@ -49,7 +49,7 @@ class JwtService(
         return when (principal) {
             is PatientPrincipal -> builder.forPatientRegistration()
             is RegisteredUserPrincipal -> builder.forRegisteredUser(principal)
-        }.sign(algorithm).let(::JwtResponseDtoOut)
+        }.sign(algorithm).let(::BearerTokenDtoOut)
     }
 
     private fun JWTCreator.Builder.forPatientRegistration() = withSubject(PUBLIC_SUBJECT)

--- a/backend/src/main/kotlin/blue/mild/covid/vaxx/dao/Answer.kt
+++ b/backend/src/main/kotlin/blue/mild/covid/vaxx/dao/Answer.kt
@@ -7,7 +7,7 @@ import java.time.Instant
 
 object Answer : Table("answers") {
     val created: Column<Instant> = timestamp("created")
-    val updated: Column<Instant?> = timestamp("updated").nullable()
+    val updated: Column<Instant> = timestamp("updated")
 
     val questionId = varchar("question_id", 36) references Question.id
     val patientId = varchar("patient_id", 36) references Patient.id

--- a/backend/src/main/kotlin/blue/mild/covid/vaxx/dao/DatabaseSetup.kt
+++ b/backend/src/main/kotlin/blue/mild/covid/vaxx/dao/DatabaseSetup.kt
@@ -16,12 +16,11 @@ object DatabaseSetup {
      */
     fun connect(dbConfiguration: DatabaseConfigurationDto) =
         Database.connect(
-            url = dbConfiguration.url,
+            url = "${dbConfiguration.url}?reWriteBatchedInserts=true",
             user = dbConfiguration.userName,
             password = dbConfiguration.password,
             driver = "org.postgresql.Driver"
         )
-
     /**
      * Returns true if the app is connected to database.
      */

--- a/backend/src/main/kotlin/blue/mild/covid/vaxx/dao/InsuranceCompany.kt
+++ b/backend/src/main/kotlin/blue/mild/covid/vaxx/dao/InsuranceCompany.kt
@@ -1,13 +1,14 @@
 package blue.mild.covid.vaxx.dao
 
 enum class InsuranceCompany(
-    val csFullName: String
+    val csFullName: String,
+    val code: Int
 ) {
-    CZPZ("Česká průmyslová zdravotní pojišťovna"),
-    OZP("Oborová zdravotní pojišťovna"),
-    RBP("RBP, zdravotní pojišťovna"),
-    VZP("Všeobecná zdravotní pojišťovna"),
-    VOZP("Vojenská Zdravotní Pojišťovna"),
-    ZPS("Zaměstnanecká pojišťovna Škoda"),
-    ZPMV("Zdravotní pojišťovna Ministerstva vnitra ČR")
+    CZPZ("Česká průmyslová zdravotní pojišťovna", 205),
+    OZP("Oborová zdravotní pojišťovna zaměstnanců bank, pojišťoven a stavebnictví", 207),
+    RBP("RBP, zdravotní pojišťovna", 213),
+    VZP("Všeobecná zdravotní pojišťovna České republiky", 111),
+    VOZP("Vojenská zdravotní pojišťovna ČR", 201),
+    ZPS("Zaměstnanecká pojišťovna Škoda", 209),
+    ZPMV("Zdravotní pojišťovna ministerstva vnitra", 211)
 }

--- a/backend/src/main/kotlin/blue/mild/covid/vaxx/dao/Patient.kt
+++ b/backend/src/main/kotlin/blue/mild/covid/vaxx/dao/Patient.kt
@@ -8,7 +8,7 @@ import java.time.Instant
 object Patient : Table("patients") {
     val id = varchar("id", 36)
     val created: Column<Instant> = timestamp("created")
-    val updated: Column<Instant?> = timestamp("updated").nullable()
+    val updated: Column<Instant> = timestamp("updated")
 
     val firstName = varchar("first_name", 256)
     val lastName = varchar("last_name", 256)

--- a/backend/src/main/kotlin/blue/mild/covid/vaxx/dao/Question.kt
+++ b/backend/src/main/kotlin/blue/mild/covid/vaxx/dao/Question.kt
@@ -1,17 +1,14 @@
 package blue.mild.covid.vaxx.dao
 
-import org.jetbrains.exposed.sql.Column
 import org.jetbrains.exposed.sql.Table
 import org.jetbrains.exposed.sql.`java-time`.timestamp
-import java.time.Instant
 
 object Question : Table("questions") {
     val id = varchar("id", 36)
-    val created: Column<Instant> = timestamp("created")
-    val updated: Column<Instant?> = timestamp("updated").nullable()
+    val created = timestamp("created")
+    val updated = timestamp("updated")
 
     val placeholder = varchar("placeholder", 256)
-
     val cs = text("cs")
     val eng = text("eng")
 

--- a/backend/src/main/kotlin/blue/mild/covid/vaxx/dto/response/BearerTokenDtoOut.kt
+++ b/backend/src/main/kotlin/blue/mild/covid/vaxx/dto/response/BearerTokenDtoOut.kt
@@ -1,0 +1,3 @@
+package blue.mild.covid.vaxx.dto.response
+
+data class BearerTokenDtoOut(val token: String)

--- a/backend/src/main/kotlin/blue/mild/covid/vaxx/dto/response/HealthDtoOut.kt
+++ b/backend/src/main/kotlin/blue/mild/covid/vaxx/dto/response/HealthDtoOut.kt
@@ -1,3 +1,0 @@
-package blue.mild.covid.vaxx.dto.response
-
-data class HealthDtoOut(val health: String)

--- a/backend/src/main/kotlin/blue/mild/covid/vaxx/dto/response/InsuranceCompanyDetailsDtoOut.kt
+++ b/backend/src/main/kotlin/blue/mild/covid/vaxx/dto/response/InsuranceCompanyDetailsDtoOut.kt
@@ -9,4 +9,5 @@ data class InsuranceCompanyDetailsDtoOut(
 ) {
     val name = insuranceCompany.name
     val csFullName = insuranceCompany.csFullName
+    val code = insuranceCompany.code
 }

--- a/backend/src/main/kotlin/blue/mild/covid/vaxx/dto/response/JwtDtoOut.kt
+++ b/backend/src/main/kotlin/blue/mild/covid/vaxx/dto/response/JwtDtoOut.kt
@@ -1,3 +1,0 @@
-package blue.mild.covid.vaxx.dto.response
-
-data class JwtResponseDtoOut(val token: String)

--- a/backend/src/main/kotlin/blue/mild/covid/vaxx/dto/response/PatientDtoOut.kt
+++ b/backend/src/main/kotlin/blue/mild/covid/vaxx/dto/response/PatientDtoOut.kt
@@ -1,14 +1,14 @@
 package blue.mild.covid.vaxx.dto.response
 
 import blue.mild.covid.vaxx.dao.InsuranceCompany
-import java.time.Instant
 import blue.mild.covid.vaxx.dto.AnswerDto
+import java.time.Instant
 import java.util.UUID
 
 data class PatientDtoOut(
     val id: UUID,
     val created: Instant,
-    val updated: Instant?,
+    val updated: Instant,
     val firstName: String,
     val lastName: String,
     val personalNumber: String,

--- a/backend/src/main/kotlin/blue/mild/covid/vaxx/dto/response/ServiceHealthDtoOut.kt
+++ b/backend/src/main/kotlin/blue/mild/covid/vaxx/dto/response/ServiceHealthDtoOut.kt
@@ -1,0 +1,3 @@
+package blue.mild.covid.vaxx.dto.response
+
+data class ServiceHealthDtoOut(val health: String)

--- a/backend/src/main/kotlin/blue/mild/covid/vaxx/routes/ServiceRoutes.kt
+++ b/backend/src/main/kotlin/blue/mild/covid/vaxx/routes/ServiceRoutes.kt
@@ -1,7 +1,7 @@
 package blue.mild.covid.vaxx.routes
 
 import blue.mild.covid.vaxx.dao.DatabaseSetup
-import blue.mild.covid.vaxx.dto.response.HealthDtoOut
+import blue.mild.covid.vaxx.dto.response.ServiceHealthDtoOut
 import blue.mild.covid.vaxx.dto.response.VersionDtoOut
 import blue.mild.covid.vaxx.extensions.di
 import com.papsign.ktor.openapigen.route.info
@@ -26,13 +26,13 @@ fun NormalOpenAPIRoute.serviceRoutes() {
 
     route(Routes.status).get<Unit, Unit> { respond(Unit) }
 
-    route(Routes.statusHealth).get<Unit, HealthDtoOut> {
+    route(Routes.statusHealth).get<Unit, ServiceHealthDtoOut> {
         if (DatabaseSetup.isConnected()) {
-            respond(HealthDtoOut("healthy"))
+            respond(ServiceHealthDtoOut("healthy"))
         } else {
             // TODO solve how to use different response code
             // https://github.com/papsign/Ktor-OpenAPI-Generator/wiki/A-few-examples
-            respond(HealthDtoOut("DB connection is not working"))
+            respond(ServiceHealthDtoOut("DB connection is not working"))
         }
     }
 }

--- a/backend/src/main/kotlin/blue/mild/covid/vaxx/routes/UserRoute.kt
+++ b/backend/src/main/kotlin/blue/mild/covid/vaxx/routes/UserRoute.kt
@@ -4,7 +4,7 @@ import blue.mild.covid.vaxx.auth.JwtService
 import blue.mild.covid.vaxx.auth.PatientPrincipal
 import blue.mild.covid.vaxx.dto.request.CaptchaVerificationDtoIn
 import blue.mild.covid.vaxx.dto.request.LoginDtoIn
-import blue.mild.covid.vaxx.dto.response.JwtResponseDtoOut
+import blue.mild.covid.vaxx.dto.response.BearerTokenDtoOut
 import blue.mild.covid.vaxx.extensions.di
 import blue.mild.covid.vaxx.service.UserLoginService
 import com.papsign.ktor.openapigen.route.info
@@ -22,7 +22,7 @@ fun NormalOpenAPIRoute.userRoutes() {
     val jwtService by di().instance<JwtService>()
 
     route(Routes.registeredUserLogin) {
-        post<Unit, JwtResponseDtoOut, LoginDtoIn>(
+        post<Unit, BearerTokenDtoOut, LoginDtoIn>(
             info("Login endpoint for the registered users such as administrators and doctors.")
         ) { _, loginDto ->
             val principal = userLoginService.verifyCredentials(loginDto.username, loginDto.password)
@@ -30,7 +30,7 @@ fun NormalOpenAPIRoute.userRoutes() {
         }
     }
     route(Routes.registrationCaptcha) {
-        post<Unit, JwtResponseDtoOut, CaptchaVerificationDtoIn>(
+        post<Unit, BearerTokenDtoOut, CaptchaVerificationDtoIn>(
             info("Endpoint that issues access to the API for the non-registered users.")
         ) { _, (_) ->
             // TODO somehow verify token with Google

--- a/backend/src/main/kotlin/blue/mild/covid/vaxx/service/EntityIdProvider.kt
+++ b/backend/src/main/kotlin/blue/mild/covid/vaxx/service/EntityIdProvider.kt
@@ -1,0 +1,7 @@
+package blue.mild.covid.vaxx.service
+
+import java.util.UUID
+
+class EntityIdProvider {
+    fun generateId(): Pair<UUID, String> = UUID.randomUUID().let { it to it.toString() }
+}

--- a/backend/src/main/kotlin/blue/mild/covid/vaxx/service/PatientService.kt
+++ b/backend/src/main/kotlin/blue/mild/covid/vaxx/service/PatientService.kt
@@ -22,7 +22,8 @@ import java.util.UUID
 
 class PatientService(
     private val validationService: ValidationService,
-    private val instantTimeProvider: InstantTimeProvider
+    private val instantTimeProvider: InstantTimeProvider,
+    private val entityIdProvider: EntityIdProvider
 ) {
 
     suspend fun getPatientById(patientId: UUID): PatientDtoOut = newSuspendedTransaction {
@@ -31,48 +32,42 @@ class PatientService(
             .select { Patient.id eq patientId.toString() }
             .toList() // eager fetch all data from the database
 
-        val answers = data
-            .map { AnswerDto(it[Answer.questionId].toUuid(), it[Answer.value]) }
+        val answers = data.map { AnswerDto(it[Answer.questionId].toUuid(), it[Answer.value]) }
 
-        data
-            .firstOrNull()
-            ?.let {
-                PatientDtoOut(
-                    id = it[Patient.id].toUuid(),
-                    firstName = it[Patient.firstName],
-                    lastName = it[Patient.lastName],
-                    personalNumber = it[Patient.personalNumber],
-                    phoneNumber = it[Patient.phoneNumber],
-                    email = it[Patient.email],
-                    insuranceCompany = it[Patient.insuranceCompany],
-                    answers = answers,
-                    created = it[Patient.created],
-                    updated = it[Patient.updated]
-                )
-            } ?: throw entityNotFound<Patient>(Patient::id, patientId)
+        data.firstOrNull()?.let {
+            PatientDtoOut(
+                id = it[Patient.id].toUuid(),
+                firstName = it[Patient.firstName],
+                lastName = it[Patient.lastName],
+                personalNumber = it[Patient.personalNumber],
+                phoneNumber = it[Patient.phoneNumber],
+                email = it[Patient.email],
+                insuranceCompany = it[Patient.insuranceCompany],
+                answers = answers,
+                created = it[Patient.created],
+                updated = it[Patient.updated]
+            )
+        } ?: throw entityNotFound<Patient>(Patient::id, patientId)
     }
 
     suspend fun getAllPatients(): List<PatientDtoOut> =
         newSuspendedTransaction { getAndMapPatients() }
 
     suspend fun getPatientsByPersonalNumber(patientPersonalNumber: String): List<PatientDtoOut> {
-        validationService.validatePersonalNumberAndThrow(patientPersonalNumber)
-        return newSuspendedTransaction { getAndMapPatients { Patient.personalNumber eq normalizePersonalNumber(patientPersonalNumber) } }
+        val normalizedPersonalNumber = normalizePersonalNumber(patientPersonalNumber)
+        return newSuspendedTransaction { getAndMapPatients { Patient.personalNumber eq normalizedPersonalNumber } }
     }
 
-    suspend fun getPatientsByEmail(email: String): List<PatientDtoOut> {
-        validationService.validateEmailAndThrow(email)
-        return newSuspendedTransaction { getAndMapPatients { Patient.email eq email } }
-    }
+    suspend fun getPatientsByEmail(email: String): List<PatientDtoOut> =
+        newSuspendedTransaction { getAndMapPatients { Patient.email eq email } }
 
     suspend fun savePatient(patientRegistrationDto: PatientRegistrationDtoIn) = newSuspendedTransaction {
         validationService.validatePatientRegistrationAndThrow(patientRegistrationDto)
 
-        val patientId: UUID = UUID.randomUUID()
-        val patientIdString = patientId.toString()
+        val (entityId, stringId) = entityIdProvider.generateId()
 
         Patient.insert {
-            it[id] = patientIdString
+            it[id] = stringId
             it[firstName] = patientRegistrationDto.firstName
             it[lastName] = patientRegistrationDto.lastName
             it[personalNumber] = normalizePersonalNumber(patientRegistrationDto.personalNumber)
@@ -81,14 +76,22 @@ class PatientService(
             it[insuranceCompany] = patientRegistrationDto.insuranceCompany
         }
 
-        Answer.batchInsert(patientRegistrationDto.answers) {
-            this[Answer.created] = instantTimeProvider.now()
-            this[Answer.patientId] = patientIdString
+        val now = instantTimeProvider.now()
+        // even though this statement prints multiple insert into statements
+        // they are in a fact translated to one thanks to reWriteBatchedInserts=true
+        // see https://github.com/JetBrains/Exposed/wiki/DSL#batch-insert
+        Answer.batchInsert(patientRegistrationDto.answers, shouldReturnGeneratedValues = false) {
+            // we want to specify these values because DB defaults don't support in batch inserts
+            // however, the real value will be set by the database once inserted
+            this[Answer.created] = now
+            this[Answer.updated] = now
+
+            this[Answer.patientId] = stringId
             this[Answer.questionId] = it.questionId.toString()
             this[Answer.value] = it.value
         }
 
-        PatientCreatedDtoOut(patientId)
+        PatientCreatedDtoOut(entityId)
     }
 
     suspend fun deletePatientById(patientId: UUID) = newSuspendedTransaction {
@@ -97,7 +100,7 @@ class PatientService(
         else throw entityNotFound<Patient>(Patient::id, patientId)
     }
 
-    private fun normalizePersonalNumber(personalNumber: String ): String = personalNumber.replace("/", "")
+    private fun normalizePersonalNumber(personalNumber: String): String = personalNumber.replace("/", "")
 
     private fun getAndMapPatients(where: (SqlExpressionBuilder.() -> Op<Boolean>)? = null) =
         Patient
@@ -105,11 +108,11 @@ class PatientService(
             .let { if (where != null) it.select(where) else it.selectAll() }
             .toList() // eager fetch all data from the database
             .let { data ->
-                val answers = data
-                    .groupBy({ it[Patient.id] }, { AnswerDto(it[Answer.questionId].toUuid(), it[Answer.value]) })
-
-                data
-                    .distinctBy { it[Patient.id] }
+                val answers = data.groupBy(
+                    { it[Patient.id] },
+                    { AnswerDto(it[Answer.questionId].toUuid(), it[Answer.value]) }
+                )
+                data.distinctBy { it[Patient.id] }
                     .map {
                         PatientDtoOut(
                             id = it[Patient.id].toUuid(),

--- a/backend/src/main/kotlin/blue/mild/covid/vaxx/setup/DependencyInjection.kt
+++ b/backend/src/main/kotlin/blue/mild/covid/vaxx/setup/DependencyInjection.kt
@@ -1,5 +1,6 @@
 package blue.mild.covid.vaxx.setup
 
+import blue.mild.covid.vaxx.service.EntityIdProvider
 import blue.mild.covid.vaxx.service.InstantTimeProvider
 import blue.mild.covid.vaxx.service.PatientService
 import blue.mild.covid.vaxx.service.QuestionService
@@ -11,9 +12,10 @@ import org.kodein.di.instance
 import org.kodein.di.singleton
 
 fun DI.MainBuilder.registerClasses() {
+    bind<EntityIdProvider>() with singleton { EntityIdProvider() }
     bind<InstantTimeProvider>() with singleton { InstantTimeProvider() }
     bind<QuestionService>() with singleton { QuestionService() }
     bind<ValidationService>() with singleton { ValidationService(instance()) }
-    bind<PatientService>() with singleton { PatientService(instance(), instance()) }
+    bind<PatientService>() with singleton { PatientService(instance(), instance(), instance()) }
     bind<UserLoginService>() with singleton { UserLoginService() }
 }

